python/pyarrow/tests/test_fs.py

```diff
@@ -1839,6 +1846,20 @@ def test_s3_real_aws_region_selection():
         's3://x-arrow-nonexistent-bucket?region=us-east-3')
     assert fs.region == 'us-east-3'
 
+    # Nonexistent bucket but "allow_delayed_open" delays failure until I/O is performed
+    fs = S3FileSystem(allow_delayed_open=True, allow_bucket_creation=False)
+    with pytest.raises(IOError, match="AWS Error NO_SUCH_BUCKET"):
+        with fs.open_output_stream("x-arrow-nonexistent-bucket/test.txt") as stream:
+            stream.write(b"test")
+
+    # Nonexistent bucket. from_uri is not delaying evaluation and raising OSError
+    # S3FileSystem.from_uri also fails
+    fs, path = FileSystem.from_uri(
+        's3://x-arrow-nonexistent-bucket/test.txt?allow_delayed_open=true')
+    with pytest.raises(IOError):
+        with fs.open_output_stream(path) as stream:
+            stream.write(b"test")
```

Comment:
@AlenkaF here is the test you asked.

One behavior I wasn't expecting is the immediate failure of (lines 1857-1858):
```python
fs, path = FileSystem.from_uri(
        's3://x-arrow-nonexistent-bucket/test.txt?allow_delayed_open=true')
```
as explained in comments this line fails because the existence of the bucket is checked when this is executed. maybe this is not what someone would expect when passing `allow_delayed_open` as a query parameter ?

I thought it would be better to check double check and in case it doesn't make sense revert the chances in `cpp/src/arrow/filesystem/s3fs.cc`. cc @pitrou

Reply:
I agree that this is not the expected behavior. We should fix how `from_uri` behaves when `allow_delayed_open` is present. We can do that as a separate issue in that case we should revert the changes on `s3fs.cc` to expose `allow_delayed_open` as an uri parameter as it is not working as expected.

Reply:
Hmm, I disagree. `allow_delayed_open` is about file opens and does not apply to the filesystem instantiation. It is a performance optimization as it allows writing small S3 files in a single S3 upload request (the default would be to use a multipart upload because we don't know up front how large the file will be). The _aim_ is not to delay errors, that's just a side effect.

Reply:
Oh! sorry, you are correct, I misunderstood, the test should use an existing bucket but a non-existent file, right? In this case the instantiation should be correct but should fail when reading the file.

Reply:
No, this currently only applies to writing a file. That is why the example of a non-existent bucket is given, but another situation would be trying to write to an existing bucket with insufficient permissions (less easy to test in unit tests, though).

Reply:
Perhaps passing an explicit `region` would bypass the bucket lookup, though.